File: src/main/java/io/github/smootheez/modrinth/ModrinthMetadata.java
package io.github.smootheez.modrinth;

import com.google.gson.annotations.*;
import io.github.smootheez.*;
import lombok.*;

import java.util.*;

/**
 * Represents the complete metadata payload required when uploading a version to Modrinth.
 *
 * @param projectId    the unique project ID on Modrinth
 * @param name         the display name of the version
 * @param versionNumber the semantic or custom version identifier
 * @param changelog    the changelog text associated with this version
 * @param gameVersions list of supported game versions (e.g. "1.20.1")
 * @param loaders      the supported mod loaders (Fabric, Forge, Quilt, etc.)
 * @param releaseChannel the release channel (release, beta, alpha)
 * @param dependencies  dependency metadata for this version
 * @param featured      whether the version should be featured
 * @param status        the Modrinth version status (listed, draft, archived, etc.)
 * @param fileParts     file identifiers used during multipart upload
 */
@Builder
public record ModrinthMetadata(

        @SerializedName("project_id")
        String projectId,

        @SerializedName("name")
        String name,

        @SerializedName("version_number")
        String versionNumber,

        @SerializedName("changelog")
        String changelog,

        @SerializedName("game_versions")
        List<String> gameVersions,

        @SerializedName("loaders")
        List<LoaderType> loaders,

        @SerializedName("release_channel")
        String releaseChannel,

        @SerializedName("dependencies")
        List<DependencyMetadata> dependencies,

        @SerializedName("featured")
        boolean featured,

        @SerializedName("status")
        String status,

        @SerializedName("file_parts")
        List<String> fileParts
) { }

File: src/main/kotlin/io/github/smootheez/modrinth/ModrinthMetadata.kt
package io.github.smootheez.modrinth

import io.github.smootheez.LoaderType
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ModrinthMetadata(
    @SerialName("project_id")
    val projectId: String,

    @SerialName("name")
    val name: String? = null,

    @SerialName("version_number")
    val versionNumber: String,

    @SerialName("changelog")
    val changelog: String,

    @SerialName("game_versions")
    val gameVersions: List<String>,

    @SerialName("loaders")
    val loaders: List<LoaderType>,

    @SerialName("release_channel")
    val releaseChannel: String, // default to RELEASE, which is the most common case

    @SerialName("dependencies")
    val dependencies: List<DependencyMetadata> = emptyList(),

    @SerialName("featured")
    val featured: Boolean,

    // a new version, Modrinth defaults status to "listed". nullable with defaults to simplify upload calls
    @SerialName("status")
    val status: String = "listed",

    @SerialName("file_parts")
    val fileParts: List<String>
)

@Serializable
data class DependencyMetadata(
    @SerialName("project_id")
    val projectId: String?,

    @SerialName("dependency_type")
    val dependencyType: DependencyType
)

